components/org.wso2.carbon.identity.branding.preference.resolver/src/main/java/org/wso2/carbon/identity/branding/preference/resolver/UIBrandingPreferenceResolverImpl.java

```diff
@@ -173,24 +186,27 @@ private BrandingPreference resolveOrganizationBranding(String name, String local
         }
 
         if (organizationId != null) {
-            BrandedOrgCacheEntry valueFromCache =
-                    brandedOrgCache.getValueFromCache(new BrandedOrgCacheKey(organizationId), currentTenantDomain);
-            if (valueFromCache != null) {
-                String brandingResolvedTenantDomain = valueFromCache.getBrandingResolvedTenant();
-                BrandingPreference resolvedBrandingPreference = getPreference(ORGANIZATION_TYPE, name, locale,
-                        brandingResolvedTenantDomain);
-
-                if (!currentTenantDomain.equals(brandingResolvedTenantDomain)) {
-                    // Since Branding is inherited from an ancestor org, removing the ancestor org displayName.
-                    removeOrgDisplayNameFromBrandingPreference(resolvedBrandingPreference);
+            Optional<BrandingPreference> resolvedBrandingPreference;
+            if (restrictToPublished) {
+                resolvedBrandingPreference = getOrganizationBrandingFromCache(name, locale,
+                        organizationId + PUBLISHED_BRANDING_CACHE_KEY_SUFFIX, currentTenantDomain);
+                if (resolvedBrandingPreference.isPresent()) {
+                    return resolvedBrandingPreference.get();
                 }
-                return resolvedBrandingPreference;
+            }
+            resolvedBrandingPreference =
+                    getOrganizationBrandingFromCache(name, locale, organizationId, currentTenantDomain);
+            if (resolvedBrandingPreference.isPresent() &&
+                    (!restrictToPublished || isBrandingPublished(resolvedBrandingPreference.get()))) {
+                return resolvedBrandingPreference.get();
             }
 
             // No cache found. Start with current organization.
             Optional<BrandingPreference> brandingPreference =
                     getBrandingPreference(ORGANIZATION_TYPE, name, locale, currentTenantDomain);
-            if (brandingPreference.isPresent()) {
+            if (brandingPreference.isPresent() &&
+                    (!restrictToPublished ||
+                            BrandingPreferenceMgtUtils.isBrandingPublished(brandingPreference.get()))) {
```

Comment:
Shall we extract this logic to a private class since there are multiple places using this condition. WDYT?

Reply:
Addressed with https://github.com/wso2-extensions/identity-branding-preference-management/pull/43/commits/af5c5fc03be0f65a78238ae5f63f68958df045d8